### PR TITLE
Update source packaging

### LINKED ISSUE #126: [BUG] Include `requirements.txt` in PyPI source
The source is missing `requirements.txt`. This results in failure of installing `unionml` from the source.

> Modifications in `MANIFEST.in` can fix this.

image

### LINKED ISSUE #127: Add `.gitattributes` file for GitHub releases.
Adding `.gitattributes` file to the repository will help in controlling what to include/exclude in GitHub releases. By default, everything in the repository is included in the releases.

### LINKED ISSUE #129: Single source version (`unionml.__version__`)
Single source version (`unionml.__version__`) from `setup.py`.

Currently `__version__` is set at two places by hardcoding the values. 

- `setup.py`
- `unionml/__init__.py`

The values are updated by the `publish.yml` workflow. Instead of updating the version in two places, it could be made possible to update the `__version__` only in `setup.py` and set the version keyword of `setup()` function. This can then be pulled by `unionml/__init__.py` so that when you run `import unionml; print(unionml.__version__)` that returns the version set by the setup.py file.

### PR DESCRIPTION
**Pull request by sugatoray**
The source is missing `requirements.txt`. This PR will ensure such slip-ups are avoided by design with minimal maintenance in future.

image

- [x] Update `MANIFEST.in`: Closes #126 
  - Explicitly specify files/folders to include or exclude from the PyPI source (`unionml-#.#.#.tar.gz` file).
  - This also helps separate the concern of what to package in the PyPI source from the rest of the logic in `setup.py` .

- [x] Add `.gitattributes` file to control what to include/exclude in GitHub releases: Closes #127

- [x] Single source version (`unionml.__version__`) from `setup.py`: Closes #129
  - [x] changes `unionml/__init__.py`
  - [x] changes workflow: `publish.yml`

### CODE REVIEW

**Comment by sugatoray** [comment]
cc: @cosmicBboy

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -33,8 +33,10 @@ jobs:
           # from refs/tags/v1.2.3 get 1.2.3
           VERSION=$(echo $GITHUB_REF | sed 's#.*/v##')
           PLACEHOLDER="__version__\ =\ \"0.0.0+dev0\""
-          grep "$PLACEHOLDER" "unionml/__init__.py"
-          sed -i "s#$PLACEHOLDER#__version__ = \"$VERSION\"#g" "unionml/__init__.py"
+          # grep "$PLACEHOLDER" "unionml/__init__.py"
+          # sed -i "s#$PLACEHOLDER#__version__ = \"$VERSION\"#g" "unionml/__init__.py"
+          grep "$PLACEHOLDER" "setup.py"
+          sed -i "s#$PLACEHOLDER#__version__ = \"$VERSION\"#g" "setup.py"
```
Comment (sugatoray) [inline .github/workflows/publish.yml:39]: Looking forward to your comments on this.

cc: @cosmicBboy

File: setup.py [hunk 2]
```diff
@@ -16,7 +18,7 @@
     description="The easiest way to build and deploy machine learning microservices.",
     long_description=long_description,
```
Comment (cosmicBboy) [inline setup.py:19]: update this to use `README`

File: .gitattributes [hunk 3]
```diff
@@ -0,0 +1,34 @@
+## NOTES
+#
+# Author: Sugato Ray (https://github.com/sugatoray/genespeak)
+# Source: https://github.com/sugatoray/genespeak/blob/c3f57ad9fbedd5845fd279adcd83826b9a32e952/.gitattributes
```
Comment (cosmicBboy) [inline .gitattributes:4]: I think we can avoid adding comments here with author annotations... we'll properly attribute you as a contributor and git will keep track of what parts you contributed 😀

File: .github/workflows/publish.yml (hunk 1 above)
Comment (cosmicBboy) [reply .github/workflows/publish.yml:39]: is the `sed` replacement in `unionml/__init__.py` still necessary after your changes?

File: MANIFEST.in [hunk 4]
```diff
@@ -1 +1,39 @@
...
+include *.cff  # citation info
+
+include MANIFEST.in
+include pyproject.toml
+include setup.py
+include setup.cfg
+include tox.ini
+
+include requirements.txt
+recursive-include requirements *.txt
```
Comment (cosmicBboy) [inline MANIFEST.in:20]: what's this recursive-include for? the project currently has no `requirements` directory.

**Review by cosmicBboy — COMMENTED** [review]
Good stuff @sugatoray ! See me comments

File: MANIFEST.in (hunk 4 above)
Comment (sugatoray) [reply MANIFEST.in:20]: I left it there for you to decide if you would like to move the dev and docs requirements into a `requirements` folder.

Comment (sugatoray) [reply MANIFEST.in:20]: > is the sed replacement in unionml/__init__.py still necessary after your changes?

No, it's not necessary. But did not want to drop it before discussing with you.

File: setup.py (hunk 2 above)
Comment (sugatoray) [reply setup.py:19]: Thanks. Good catch!

File: .github/workflows/publish.yml (hunk 1 above)
Comment (sugatoray) [reply .github/workflows/publish.yml:39]: Made changes now to to only use `sed` once to apply to `setup.py`.

**Comment by sugatoray** [comment]
@cosmicBboy I have made the changes. Please see and let me know if you think anything else must be addressed before merging.

File: setup.py (hunk 2 above)
Comment (sugatoray) [reply setup.py]: Done.

**Comment by cosmicBboy** [comment]
there's a mypy issue: https://github.com/unionai-oss/unionml/runs/7010795478?check_suite_focus=true#step:6:63

gonna add a `# type: ignore` comment

**Comment by cosmicBboy** [comment]
@sugatoray looks like there are some linting/formatting errors: please follow the contribution guide to run the `pre-commit` hooks:

https://unionml.readthedocs.io/en/latest/contributing_guide.html#pre-commit-setup

**Comment by sugatoray** [comment]
I was pushing these changes from the web version of VSCode as an experiment to see what kind of challenges I might come across if only using VS Code to push code changes. 

Will now clone locally and apply pre-commit as well. Thank you, for pointing it out.

**Comment by sugatoray** [comment]
## Known issue with `pre-commit` + VS Code `git` integration

I use VS Code. If you are using **VS Code + `pre-commit` + conda**, you may see some weird behavior in VS Code, while using the `git` integration/extension it provides out of the box for committing files. I opened an issue here.

- https://github.com/microsoft/vscode/issues/153037

**Comment by sugatoray** [comment]
Added constrained installation of `importlib-metadata` only for python version `<3.8`.

Source: https://github.com/huggingface/huggingface_hub/blob/66175504f943f080e0b407e04cf56bd0895f6bb9/setup.py#L20

```diff
- importlib-metadata
+ importlib-metadata;python_version<'3.8'
```

**Comment by cosmicBboy** [comment]
✅ LGTM! Merging now... congrats on your first contribution @sugatoray ! 🎉🎉🎉

**Comment by sugatoray** [comment]
@cosmicBboy Thank you for the reviews and merging. :rocket:

**Pull request merged**